Mesh Documentation (#4504)
SciTools/iris: docs/src/further_topics/ugrid/data_model.rst

.. include:: ../../common_links.inc

> .. _ugrid model:

Reviewer: Looking good.
But overall I wish we could say 'unstructured data' a bit more, and 'UGRID' a bit less.
The way I see it, we have borrowed UGRID as a template for our data model, enabling us to represent unstructured data.  But we aim to support unstructured data, per se, whatever form we got it in.  
So we want to make the point that this is (or at least will be) a general facility in Iris.
We kind-of proved that _already_, by interoperating with ESMF grids in esmf-regrid and VTK data in Geovista.

Author: 77e3f95aa4bf4b57e1c60788fd0462191e6645ce

.. include:: ../../common_links.inc

.. _ugrid model:

The Mesh Data Model
*******************

.. important::

        This page is intended to summarise the essentials that Iris users need
        to know about meshes. For exhaustive details on UGRID itself:
        `visit the official UGRID conventions site`__.

Evolution, not revolution
=========================
Mesh support has been designed wherever possible to fit within the existing
Iris model. Meshes concern only the spatial location of data, and can
optionally be limited to just the horizontal location (e.g. X and Y). Other

> dimensions such as time or experimental re-runs (and often vertical levels)

Reviewer 2: ```suggestion
dimensions such as time or ensemble member (and often vertical levels)
```

Reviewer 2: I found the "suggest changes" button! Most will be super picky/pedantic - feel free to ignore.

Author: 1d94cc892fcefdc2da2abdd32e593d1fb659d3b8

retain their familiar structured format.

The UGRID conventions themselves are designed as an addition to the existing CF
conventions, which are at the core of Iris' philosophy.

What's Different?
=================

The mesh format represents data's geographic location using an **unstructured
mesh**. This has significant pros and cons when compared to a structured grid.

.. contents::
   :local:

The Detail
----------
..
    The diagram images are SVG's, so editable by any graphical software
     (e.g. Inkscape). They were originally made in MS PowerPoint.

    Uses the IBM Colour Blind Palette (see
     http://ibm-design-language.eu-de.mybluemix.net/design/language/resources/color-library
     )

Structured Grids
~~~~~~~~~~~~~~~~
Assigning data to locations using a structured grid is essentially an act of
matching coordinate arrays to each dimension of the data array. The data can
also be represented as an area (instead of a point) by including a bounds array
for each coordinate array.

.. figure:: images/data_structured_grid.svg
   :alt: Diagram of how data is represented on a structured grid

>    Data on a structured grid

Reviewer 2: Are there references in the text pointing to these images? I'm not sure I fully follow what the circles and crosses mean, e.g. are there x-y axes on the cross diagrams, but index axes on the dot diagrams? And the diamonds are scalar values mapped to those coords?

![image](https://user-images.githubusercontent.com/22805/150578762-a36d6b0e-4b9d-48f9-bf45-8092dd5b11de.png)

Author: I've been trying to keep the prose and diagrams self-contained - for accessibility and ease of maintenance.

And I had also hoped that the diagrams would be self-describing where possible. But from yours and others' feedback there are parts that aren't as obvious as I had thought! I'll see what I can think of to make things clearer within the diagrams, while avoiding information overload.

Author: >But from yours and others' feedback there are parts that aren't as obvious as I had thought! I'll see what I can think of to make things clearer within the diagrams, while avoiding information overload.

Other opinions on this would be gratefully received @SciTools/developers

Reviewer 3: Is the vague highlighting of the top-left point intentional?
I think I'd prefer it wasn't highlighted, and just one in the middle was? I'd also add some form of background colouring so that it's clearer what's the data, what's the objects that are part of the cube and what's just conceptual process.

Reviewer 3: Ah, I see now that it was. Can we make the two highlight types more distinct?

Author: I've had a go at addressing both:

* The distinction between arrays and geography - by including an Africa outline to indicate geography.
* The highlighting of two distinct points - bolder dash markings for the 'secondary' point.

3c0e5f85bff8e008226dc9d8a88da47b9b6f4db1

Reviewer 4: Considering there is a lot going on in this, I think it would be quite difficult for a user to follow this diagram without a bit of prose explaining
 what is going on. Maybe include something like the following in the figure caption:

> 1 dimensional Coordinate Arrays, x and y (pink circles)  are combined to construct a structured grid of points (pink crosses). If bounds arrays (blue circles) are present, the x and y bounds construct 4-sided cells (outlined with blue). The data array (orange circles) assigns data with the points (orange diamonds) if no bounds are present) or with the bounded cells (orange quadrialterals) when bounds exist. The above example highlighs the location (x=25, y=20) which has bounds (x=[23,28], y=[19,21]) 

I would also think the other figures would benefit from a more detailed caption

Reviewer 2: > I've been trying to keep the prose and diagrams self-contained - for accessibility and ease of maintenance.
> 
> And I had also hoped that the diagrams would be self-describing where possible. But from yours and others' feedback there are parts that aren't as obvious as I had thought! I'll see what I can think of to make things clearer within the diagrams, while avoiding information overload.

I understand that rationale, makes a lot of sense so feel free to ignore this advice if you feel those concerns trump this.

I know this isn't a paper or publication, but I was taught (and I think it is a good general rule of thumb) that all figures in a document should be referenced from the text, otherwise what is the context, why are they there?
A brief search for a guide to this I found the APA style guide, see https://guides.lib.monash.edu/citing-referencing/apa-tables-figures:

> * All Tables and Figures must be referred to in the main body of the text.

Author: We're gonna go with some more explicit captions on each figure, and I'll think about some cross-referencing with the text blocks.

Author: 3692c1cfa51f15181437b22088f3f223d0037a93

Author: Also I hadn't spotted that the example coordinates were unbelievable in a couple of ways.

67b1eb3765829eb70b5d4f2e581a8b8dde09b498

   :download:`full size image <images/data_structured_grid.svg>`

Unstructured Meshes
~~~~~~~~~~~~~~~~~~~
The most basic element in a mesh is the 0-dimensional **node**: a single
location in space. Every node in the mesh is defined by indexing the
1-dimensional X and Y (and optionally Z) coordinate arrays (the
``node_coordinates``) - e.g. ``(x[3], y[3])`` gives the position of the fourth
node. Note that this means each node has its own coordinates, independent of
every other node. Since nodes can be anywhere in the space - **unstructured** -

> the position in the array has nothing to do with spatial position.

Reviewer 2: ```suggestion
the position in the array (pink dot in Fig 25) has nothing to do with spatial position (corresponding pink cross in Fig 25).
```

Author: https://github.com/SciTools/iris/pull/4504#discussion_r790679955

Author: 3692c1cfa51f15181437b22088f3f223d0037a93

~~~~~~~~~~~~~~~~
Assigning data to locations using a structured grid is essentially an act of
matching coordinate arrays to each dimension of the data array. The data can
also be represented as an area (instead of a point) by including a bounds array
for each coordinate array.

.. figure:: images/data_structured_grid.svg
   :alt: Diagram of how data is represented on a structured grid

   Data on a structured grid

   :download:`full size image <images/data_structured_grid.svg>`

Unstructured Meshes
~~~~~~~~~~~~~~~~~~~
The most basic element in a mesh is the 0-dimensional **node**: a single
location in space. Every node in the mesh is defined by indexing the
1-dimensional X and Y (and optionally Z) coordinate arrays (the
``node_coordinates``) - e.g. ``(x[3], y[3])`` gives the position of the fourth
node. Note that this means each node has its own coordinates, independent of
every other node. Since nodes can be anywhere in the space - **unstructured** -
the position in the array has nothing to do with spatial position.

If data is assigned to node location it must be stored in a 1D array of equal
length to the number of nodes in the mesh. ``data[3]`` is at the position:

> ``(x[3], y[3])``.

Reviewer: This sentence starting with "data[3]" seemed to me to appear somewhat 'out of the blue', as a bit of a shock.
Possibly all it needs is a "Thus, for example... " before it ?

Author: a0b405c3b4432e6e172616620cb6a1563b083a57

* | :attr:`~iris.cube.Cube.location`
  | ``node``/``edge``/``face`` - the mesh element type with which this
    :class:`~iris.cube.Cube`\'s :attr:`~iris.cube.Cube.data` is associated.
* | :meth:`~iris.cube.Cube.mesh_dim`
  | The :class:`~iris.cube.Cube`\'s **unstructured dimension** - the one that
    indexes over the horizontal :attr:`~iris.cube.Cube.data` positions.

These members will all be ``None`` for a :class:`~iris.cube.Cube` with no
associated :class:`~iris.experimental.ugrid.Mesh`.

This :class:`~iris.cube.Cube`\'s unstructured dimension has multiple attached
:class:`iris.experimental.ugrid.MeshCoord`\s (one for each axis e.g.
``x``/``y``), which can be used to infer the points and bounds of any index on
the :class:`~iris.cube.Cube`\'s unstructured dimension.

.. todo: Cube printout

The Detail
----------
How UGRID information is stored
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. todo: Mesh printout?

* | :class:`iris.experimental.ugrid.Mesh`
  | Contains all information about the mesh.
  | Includes:

  * 1-3 collections of :class:`iris.coords.AuxCoord`\s:

    * | **Required**: :attr:`~iris.experimental.ugrid.Mesh.node_coords`
      | The nodes that are the basis for the mesh.

>     * | Optional: :attr:`~iris.experimental.ugrid.Mesh.edge_coords`,

Reviewer: It may be worth a special note that we still ***require*** these at present (a specific limitation of the current Mesh implementation)

Author: a0b405c3b4432e6e172616620cb6a1563b083a57

attaching to the :class:`~iris.cube.Cube`\'s unstructured dimension, in the
same way that all :class:`~iris.coords.Coord`\s attach to
:class:`~iris.cube.Cube` dimensions. This allows a single
:class:`~iris.cube.Cube` to have a combination of unstructured and structured
dimensions (e.g. horizontal mesh plus vertical levels and a time series),
using the same logic for every dimension.

:class:`~iris.experimental.ugrid.MeshCoord`\s are instantiated using a given
:class:`~iris.experimental.ugrid.Mesh`, ``location``
("node"/"edge"/"face") and ``axis``. The process interprets the
:class:`~iris.experimental.ugrid.Mesh`\'s
:attr:`~iris.experimental.ugrid.Mesh.node_coords` and if appropriate the
:attr:`~iris.experimental.ugrid.Mesh.edge_node_connectivity`/
:attr:`~iris.experimental.ugrid.Mesh.face_node_connectivity` and
:attr:`~iris.experimental.ugrid.Mesh.edge_coords`/
:attr:`~iris.experimental.ugrid.Mesh.face_coords`
to produce a :class:`~iris.coords.Coord`
:attr:`~iris.coords.Coord.points` and :attr:`~iris.coords.Coord.bounds`
representation of all the :class:`~iris.experimental.ugrid.Mesh`\'s
nodes/edges/faces for the given axis.

> A :class:`~iris.experimental.ugrid.Mesh`

Reviewer: Reads a bit awkwardly.  I think "The method :meth:`~iris.experimental.ugrid.Mesh.to_MeshCoords` is available ... *given* only the ``location`` argument."

Author: a0b405c3b4432e6e172616620cb6a1563b083a57

        to know about meshes. For exhaustive details on UGRID itself:
        `visit the official UGRID conventions site`__.

Evolution, not revolution
=========================
Mesh support has been designed wherever possible to fit within the existing
Iris model. Meshes concern only the spatial location of data, and can
optionally be limited to just the horizontal location (e.g. X and Y). Other
dimensions such as time or ensemble member (and often vertical levels)
retain their familiar structured format.

The UGRID conventions themselves are designed as an addition to the existing CF
conventions, which are at the core of Iris' philosophy.

What's Different?
=================

The mesh format represents data's geographic location using an **unstructured
mesh**. This has significant pros and cons when compared to a structured grid.

.. contents::
   :local:

The Detail
----------
..
    The diagram images are SVG's, so editable by any graphical software
     (e.g. Inkscape). They were originally made in MS PowerPoint.

    Uses the IBM Colour Blind Palette (see
     http://ibm-design-language.eu-de.mybluemix.net/design/language/resources/color-library
     )

> Structured Grids

Reviewer 3: I'd appreciate this having a subtitle or something, e.g. "Structured Grids (the old world)" to contrast with "Unstructured meshes (the new world)" so that I know I'm not expecting surprises in the structured grids bit (given I'm not familiar with referring to them as such - to me they're just how the Iris world works).

Author: 3c0e5f85bff8e008226dc9d8a88da47b9b6f4db1

node. Note that this means each node has its own coordinates, independent of
every other node. Since nodes can be anywhere in the space - **unstructured** -
the position in the array has nothing to do with spatial position.

If data is assigned to node location it must be stored in a 1D array of equal
length to the number of nodes in the mesh. ``data[3]`` is at the position:
``(x[3], y[3])``.

Data can also be assigned to higher dimensional elements - **edges**, **faces**
or **volumes**. These elements are constructed by connecting nodes together
using a 2-dimensional **connectivity** array. One dimension varies over each
element, while the other dimension varies over each node that makes up that
element; the values in the array are the node indices. E.g. we could make 2
square faces from 6 nodes using this ``face_node_connectivity``:
``[[0, 1, 3, 2], [2, 3, 5, 4]]``. Remember that meshes are **unstructured**, so
there is no significance to the order of the faces in the array. Data assigned
to a higher dimensional location must be stored in a 1D array of equal length
to the number of elements in the mesh, e.g. ``my_data = [0.33, 4.02]`` for our

> example.

Reviewer 3: I was tracking well through this page, then this paragraph completely stumped me - the concept of "node" and "element" weren't solid enough yet that I could work out what was happening.
If it's rows and columns, can we refer to rows and columns? E.g. "Each row describes an element (edge, face or volume), by giving the nodes (points) that make it up." I think that "we could make 2 square faces, each from 4 nodes" would be much easier to understand for the example too, because otherwise I have to work out where the 6 maps to, which isn't immediately obvious.

On "Remember", I'd split to a new paragraph so that I can have a break as a user by being satisfied that I got the previous concept. I'd then reword to something like

> Data assigned to the elements described by a **connectivity** array must be stored in a 1D array with one datum per element, e.g. ``my_data = [0.33, 4.02]`` for our example because we have two square elements. Remember that meshes are **unstructured**, so there is no significance to the order of the faces in the array.

Author: 3c0e5f85bff8e008226dc9d8a88da47b9b6f4db1

.. note::

        Connectivities also exist to connect the higher dimensional elements,

>         e.g. ``face_edge_connectivity``. These are optional conveniences to

Reviewer 3: Do we not hit problems where only having node-face connectivity without node-edge and edge-face means that we define faces ambiguously because we could reorder the nodes used? Or does order in the connectivity array define that for us? Also, how do we define "inside"?

Author: >Or even that the face coords always lie strictly within the face ??

Yes I had that same moment of pause while writing the docs 😅

Author: I can address enough of the uncertainty here by mentioning the anti-clockwise thing.

        speed up certain operations and will not be discussed here.

.. figure:: images/data_ugrid_mesh.svg
   :alt: Diagram of how data is represented on an unstructured mesh

   Data on an Unstructured Mesh

   :download:`full size image <images/data_ugrid_mesh.svg>`

----

The mesh model also supports edges/faces/volumes having associated 'centre'
coordinates - to allow point data to be assigned to these elements. 'Centre' is
just a convenience term - the points can exist anywhere within their respective
elements.

> .. figure:: images/ugrid_element_centres.svg

Reviewer 3: I'd find it easier with a key here that tells me those are the "centres" and the "nodes". Also somewhere, maybe in the text, link those centres to the "face_coordinates" so I don't have to count to work out what's linked to what?

Author: >Also somewhere, maybe in the text, link those centres to the "face_coordinates" so I don't have to count to work out what's linked to what?

Aren't the labelled array sizes and the name `face_coordinates` enough for that?

Reviewer 3: I thought from `face_coordinates` that I'd be getting the corners, not the centre. I also initially thought you'd made an error only having 12 not 15, until I worked it out.

If you're worried it'll clog the image or add confusion then it's probably fine to leave out, as it can be worked out, but it took some spare brain capacity to do so which meant I couldn't use it all on the overall concepts.

Author: That's a tricky compromise. I have indeed been trying to keep the diagrams sparse - both for clarity and also to require readers to 'engage' with what they are seeing - but whether that's sensible is another matter...

Author: Have put simple labels against one face coordinate and one node coordinate.

3c0e5f85bff8e008226dc9d8a88da47b9b6f4db1

   :alt: Diagram demonstrating mesh face-centred data.

   Data can be assigned to mesh edge/face/volume 'centres'

> Mesh Flexibility

Reviewer 3: This feels different to the sections above - we're out of "what's going on" and into "why is it better". Maybe this moves into "What this means"?

Author: Are you thinking it would be better to rename this section to `Why use Meshes?` and promote it to the higher heading level?

Author: Scratch that, it definitely belongs in a subheading concerning `The Detail`. I never meant to imply that `The Detail` should be limited to "what's going on" - I just wanted to isolate the deep explanation so that those who want the 'cliff notes' can skip it.

Author: 3c0e5f85bff8e008226dc9d8a88da47b9b6f4db1

space. We must instead perform specialised operations using the information in
the connectivities present, or by translating the mesh into a format designed
for mesh analysis such as VTK.

Such calculations can still be optimised to avoid them slowing workflows, but
the important take-away here is that **adaptation is needed when working mesh
data**.

How Iris Represents This
========================

..
    Include API links to the various classes

    Include Cube/Mesh printout(s)

.. seealso::

        Remember this is a prose summary. Precise documentation is at:
        :mod:`iris.experimental.ugrid`.

.. note::

        At time of writing (``Jan 2022``), neither 3D meshes nor 3D elements
        (volumes) are supported.

The Basics
----------
The Iris :class:`~iris.cube.Cube` has several new members:

* | :attr:`~iris.cube.Cube.mesh`
  | The :class:`iris.experimental.ugrid.Mesh` that describes the
    :class:`~iris.cube.Cube`\'s horizontal geography.
* | :attr:`~iris.cube.Cube.location`
  | ``node``/``edge``/``face`` - the mesh element type with which this
    :class:`~iris.cube.Cube`\'s :attr:`~iris.cube.Cube.data` is associated.
* | :meth:`~iris.cube.Cube.mesh_dim`

>   | The :class:`~iris.cube.Cube`\'s **unstructured dimension** - the one that

Reviewer 3: This is the first use of the phrase "unstructured dimension" and is near the bottom of the page. Can we define this earlier (it might make some of the explanation easier to define this early then use it, and will make it less unexpected here)?

Author: 3c0e5f85bff8e008226dc9d8a88da47b9b6f4db1

Reviewer 3: > represented by either the coordinate or connectivity arrays detailed above

This bit of the `important` block doesn't make sense to me. My attempt to make this make sense below:

Elements are enumerated along a single unstructured dimension (present in the coordinate or connectivity arrays detailed above). An element’s position along this dimension has nothing to do with its spatial position.

Author: a31546293bde29313aa7d24dd647ac27476c6604

``faces[3]`` if it were face-located.

.. figure:: images/data_ugrid_mesh.svg
   :alt: Diagram of how data is represented on an unstructured mesh

   Data on an Unstructured Mesh

   :download:`full size image <images/data_ugrid_mesh.svg>`

----

The mesh model also supports edges/faces/volumes having associated 'centre'
coordinates - to allow point data to be assigned to these elements. 'Centre' is
just a convenience term - the points can exist anywhere within their respective
elements.

.. figure:: images/ugrid_element_centres.svg
   :alt: Diagram demonstrating mesh face-centred data.

   Data can be assigned to mesh edge/face/volume 'centres'

Mesh Flexibility
++++++++++++++++
Above we have seen how one could replicate data on a structured grid using
a mesh instead. But the utility of a mesh is the extra flexibility it offers.
Here are the main examples:

* Every node is completely independent - every one can have unique X and
  Y (and Z) coordinate values.

.. figure:: images/ugrid_node_independence.svg
   :alt: Diagram demonstrating the independence of each mesh node
   :align: center
   :width: 300

>    Every mesh node is completely independent

Reviewer 4: I don't understand what this figure is intending to represent. Particularly the arrows. Are they to depicting the transition from old world (structured grids) to new world (unstructured grids)
![image](https://user-images.githubusercontent.com/8101163/151149891-149f9a76-9879-4bde-82f7-dc37d573502b.png)

Author: I wanted to introduce the new concepts gradually. So the initial introduction of the 'new world' was replicating the 'old world' example. I wanted to demonstrate that arrays of the **_same shape and size_** could also represent something unstructured.

I hope you'll agree that's an important step. But presumably there's a better way to represent this (or maybe just including the explanation I've just given you)?

Author: https://github.com/SciTools/iris/pull/4504#discussion_r793486314

Author: https://github.com/SciTools/iris/pull/4504/commits/3692c1cfa51f15181437b22088f3f223d0037a93

For example: representing a cubed-sphere using a mesh leads to coordinates and
connectivities being **~8 times larger than the data itself**, as opposed to a
small fraction of the data size when dividing a sphere using a structured grid.

This further increases the emphasis on lazy loading and processing of data
using packages such as Dask.

.. note::

        The large, 1D data arrays associated with meshes are a very different
        shape to what Iris users and developers are used to. It is suspected
        that optimal performance will need new chunking strategies, but at time
        of writing (``Jan 2022``) experience is still limited.

.. todo:
        Revisit when we have more information.

Spatial operations on mesh data are more complex
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Detail: :doc:`operations`

Indexing a mesh data array cannot be used for:

#. Region selection
#. Neighbour identification

This is because - unlike with a structured data array - relative position in
a mesh's 1-dimensional data arrays has no relation to relative position in
space. We must instead perform specialised operations using the information in

> the connectivities present, or by translating the mesh into a format designed

Reviewer 4: I may be reading this incorrectly, but I would have thought this should be:

> We must instead perform specialised operations using the information **present** in the connectivities

Author: I was actually thinking that if there were more connectivities than the bare minimum (e.g. `face_edge_connectivity`), it would be possible to write something that quickly finds neighbours or maybe identifies regions.

This will probably do:

```suggestion
the mesh's connectivities, or by translating the mesh into a format designed
```

        :mod:`iris.experimental.ugrid`.

.. note::

        At time of writing (``Jan 2022``), neither 3D meshes nor 3D elements
        (volumes) are supported.

The Basics
----------
The Iris :class:`~iris.cube.Cube` has several new members:

* | :attr:`~iris.cube.Cube.mesh`
  | The :class:`iris.experimental.ugrid.Mesh` that describes the
    :class:`~iris.cube.Cube`\'s horizontal geography.
* | :attr:`~iris.cube.Cube.location`
  | ``node``/``edge``/``face`` - the mesh element type with which this
    :class:`~iris.cube.Cube`\'s :attr:`~iris.cube.Cube.data` is associated.
* | :meth:`~iris.cube.Cube.mesh_dim`
  | The :class:`~iris.cube.Cube`\'s **unstructured dimension** - the one that
    indexes over the horizontal :attr:`~iris.cube.Cube.data` positions.

These members will all be ``None`` for a :class:`~iris.cube.Cube` with no
associated :class:`~iris.experimental.ugrid.Mesh`.

This :class:`~iris.cube.Cube`\'s unstructured dimension has multiple attached
:class:`iris.experimental.ugrid.MeshCoord`\s (one for each axis e.g.
``x``/``y``), which can be used to infer the points and bounds of any index on
the :class:`~iris.cube.Cube`\'s unstructured dimension.

> .. todo: Cube printout

Reviewer 4: This was the one thing I felt was missing from this page so this will be a great addition

Author: 3692c1cfa51f15181437b22088f3f223d0037a93

    can be assigned to these node locations (orange diamonds) or face locations
    (orange quads) by matching the indices in the data array to the indices in
    the coordinate arrays (for nodes) or connectivity array (for faces). See
    the black outlined shapes as examples of index matching in action, and the
    black stippled shapes to demonstrate that relative array position confers
    no relative spatial information.

----

The mesh model also supports edges/faces/volumes having associated 'centre'
coordinates - to allow point data to be assigned to these elements. 'Centre' is
just a convenience term - the points can exist anywhere within their respective
elements. See :numref:`ugrid_element_centres` for a visualised example.

..  _ugrid_element_centres:
..  figure:: images/ugrid_element_centres.svg
    :alt: Diagram demonstrating mesh face-centred data.
    :align: right
    :width: 1280

    Data can be assigned to mesh edge/face/volume 'centres'

    1D *node* coordinate arrays (pink circles) describe node positions in
    space (pink crosses). A 2D connectivity array (blue circles) describes
    faces by connecting four nodes into a face outline (blue outlines on the

>     map). Further 1D *face* coordinate arrays (pink circles) describe a

Reviewer: Sorry to be late to the table here, but I think it seems unnecessarily confusing here to use the same colour for face coordinates as you do for node coordinates.  Is there a particular logic to the colour coding, which I've missed, which forces that choice ?

Author: 1. Using more than 3 colours from the [IBM colour-blind palette](https://davidmathlogic.com/colorblind/#%23648FFF-%23785EF0-%23DC267F-%23FE6100-%23FFB000) puts them too close together under some types of colour-blindness.
2. Although number 1 was my original reason, I also like that they're both coordinates and therefore are more similar to each other than the other concepts in the diagrams.

Author: There are perhaps other palettes that allow four distinct colours but I wanted to pick one that was well recognised.

    nodes (pink crosses) to make up an edge (blue lines). Data can be assigned
    to the edges (orange lines) by matching the indices of the 1D data array
    (not shown) to the indices in the connectivity array.

.. _ugrid implications:

What does this mean?
--------------------
Meshes can represent much more varied spatial arrangements
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The highly specific way of recording position (geometry) and shape
(topology) allows meshes to represent essentially **any** spatial arrangement
of data. There are therefore many new applications that aren't possible using a
structured grid, including:

* `The UK Met Office's LFRic cubed-sphere <https://hps.vi4io.org/_media/events/2018/sig-io-uk-adams.pdf>`_
* `Oceanic model outputs <https://doi.org/10.3390/jmse2010194>`_

.. todo:
        a third example!

Mesh 'payload' is much larger than with structured grids
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Coordinates are recorded per-node, and connectivities are recorded per-element.
This is opposed to a structured grid, where a single coordinate value is shared
by every data point/area along that line.

> For example: representing a cubed-sphere using a mesh leads to coordinates and

Reviewer 5: I think it's worth noting here that this is true when you are representing a cubed-sphere *with no other coordinates*.

Author: 3d18be7cdcd8937d45bec9fd83046b2c8dc497d0

                node coordinates
                    <AuxCoord: longitude / (degrees_east)  [...]  shape(5,)>
                    <AuxCoord: latitude / (degrees_north)  [...]  shape(5,)>
            edge
                edge_dimension: 'Mesh2d_edge'
                edge_node_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(6, 2)>
                edge coordinates
                    <AuxCoord: longitude / (degrees_east)  [...]  shape(6,)>
                    <AuxCoord: latitude / (degrees_north)  [...]  shape(6,)>
            face
                face_dimension: 'Mesh2d_face'
                face_node_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(2, 4)>
                face coordinates
                    <AuxCoord: longitude / (degrees_east)  [...]  shape(2,)>
                    <AuxCoord: latitude / (degrees_north)  [...]  shape(2,)>
            long_name: 'my_mesh'

* | :class:`iris.experimental.ugrid.MeshCoord`
  | Described in detail in `MeshCoords`_.
  | Stores the following information:

    * | :attr:`~iris.experimental.ugrid.MeshCoord.mesh`
      | The :class:`~iris.experimental.ugrid.Mesh` associated with this

>         :class:`~iris.experimental.ugrid.MeshCoord`. Mirrored by the

Reviewer: Might it be clearer to say
In place of "Mirrored by " ... the  /attr/ attribute of any Cube this is attached to.
Could have
 "This becomes"... or "This defines" ... or "This determines"

Author: 8bc8c8488f912d10662f26edf94bff6c315df0e0
